feat(rust): change tcp protocol serialization

SanjoDeundiak opened the pull request:
1. Optimize `Address` CBOR implementation to use byte strings instead of array of bytes which saves size BREAKING CHANGE
2. Create a `TcpTransportMessage` type dedicated to exchange messages using TCP transport exclusively with optimized encoding and support for larger messages (current max allowed size is 16 MB) BREAKING CHANGE
3. Optimize Secure Channel message encoding (remove bare wrapping) + affected by change `1.` BREAKING CHANGE
4. Add Secure Channel padding
5. Remove `protocol_version` from `LocalMessage`, `RelayMessage` and `Context`

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/kafka/protocol_aware/mod.rs:
```diff
@@ -76,11 +76,10 @@ impl KafkaMessageInterceptor for InletInterceptorImpl {
 #[rustfmt::skip]
-#[cbor(map)]
 /// Wraps the content within every record batch
 struct MessageWrapper {
```
@davide-baldo does that need a change on the elixir side?

etorreborre commented on implementations/rust/ockam/ockam_core/src/routing/address.rs:
```diff
@@ -139,6 +147,11 @@ impl Address {
     }
+
+    /// Take inner Vec
+    pub fn inner(self) -> Vec<u8> {
```
Instead of exposing the implementation, I would add a `to_local` method and use it [here](https://github.com/build-trust/ockam/pull/8164/files#diff-fb6b2451268d57544e5b8570ee3d4cff294dab6325a3f48426547f219d83c452R145).

SanjoDeundiak replied:
I'm becoming a fan of exposing implementation lately to be honest. In context of rust it's a) allows to reduce number of clones in some places b) helps borrow checker figure out when we move the inner field, and not the whole struct. That being said I would even consider making fields public more often... the only thing I don't like, is that we also give mutable access to the field as a side-effect.
Speaking of this specific case, I don't think converting the Address to local makes any sense, it's only case - test code. So I vote against making that a part of public API

etorreborre replied:
I'm definitely not a fan of exposing implementations, especially not as a blanket rule. Performance optimisations can sometimes require it, but I'd rather have this trade-off on a case-by-case basis.

etorreborre commented on implementations/rust/ockam/ockam_core/src/routing/route.rs:
```diff
@@ -252,6 +252,11 @@ impl Route {
         self.iter().all(|a| a.is_local())
     }
+
+    /// Return inner VecDeque
```
This function is not used.

SanjoDeundiak replied:
I that a problem?

etorreborre replied:
That's more code to inspect when trying to understand what's going on. But I'd rather not expose implementation details which you seem to like, seeing your comment above :-).

etorreborre commented on implementations/rust/ockam/ockam_core/src/routing/transport_type.rs:
```diff
@@ -19,6 +31,11 @@ impl TransportType {
         TransportType(n)
     }
 
+    /// Raw value
```
This function is not used.

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs:
```diff
@@ -312,17 +284,35 @@ impl EncryptorWorker {
+
+    fn add_padding(msg: SecureChannelMessage) -> SecureChannelPaddedMessage {
+        // Naїve padding of 0 to 255 zeros
+        // let padding_length: u8 = ockam_core::compat::rand::random();
```
Why is the padding commented out for now? Would that be possible to add a test to show that messages are effectively padded, to make sure that we don't have regressions there?

SanjoDeundiak replied:
We should pick the strategy for padding, which I think we'll do later. For now I only added a field so that we can add padding without any compatibility issues later. Adding a test is problematic since the API for padding is not exposed

etorreborre replied:
Can't we test something that when we encrypt the same message twice, the on-wire bytes are likely to have a different size?

SanjoDeundiak replied:
Why would it be different if we don't add any padding at the moment?

etorreborre replied:
I meant when padding is used of course :-). I was responding to your comment:
> Adding a test is problematic since the API for padding is not exposed

SanjoDeundiak replied:
Yes, so currently padding is always empty and you can't use it. So it's hard to test

etorreborre commented on implementations/rust/ockam/ockam_transport_tcp/src/protocol_version.rs:
```diff
@@ -0,0 +1,26 @@
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(TcpProtocolVersion::V1),
+            _ => Err(TransportError::InvalidProtocolVersion)?,
```
It would be good to include the incorrect version in the exception for debugging.

SanjoDeundiak replied:
I didn't want to introduce additional data to `TransportError` (currently none of its members have it)

etorreborre replied:
I think that this kind of information can make the difference when debugging production issues. Because the actual value can give you clues about what when wrong.

etorreborre commented on implementations/rust/ockam/ockam_transport_tcp/src/transport_message.rs:
```diff
@@ -0,0 +1,120 @@
+/// TCP transport message type.
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode, CborLen)]
+#[rustfmt::skip]
+pub struct TcpTransportMessage<'a> {
```
It seems to be that most of this code could stay in the `ockam_transport` crate, otherwise we will have to duplicate it for every transport.

It looks like the specificity of TCP is [here](https://github.com/build-trust/ockam/pull/8164/files#diff-8ac45ee8284813729705924bd44521c87c21eb8eeb29fd3385b8bd2ba4d9561cR122), where we serialize the data but that is independent from the `TransportMessage` datatype.

SanjoDeundiak replied:
I think it's fine to duplicate, different transports' on-the-wire structs are completely decoupled from each other. At the same time I would be OK sharing those structs, if they were exactly the same, which is not the case here. `TcpTransportMessage` doesn't have protocol type and uses CBOR only, while TransportMessage uses only BARE. It would be easier in my opinion to make further changes not worrying to break something in other transports

etorreborre replied:
I don't see why we can't have different serialization formats for the same, or mostly the same, data types. Now the code above is duplicating [some tracing logic](https://github.com/build-trust/ockam/pull/8164/files#diff-b540e777cb15df1fe532d0c6946f4934d3f201c726d6d0eb5bd2b0ee3ce30dc1R52) that has nothing to do with the serialization.

SanjoDeundiak replied:
That's a good point, do you think we should move that code to a common place and reuse?

etorreborre replied:
Yes I'd like that.

etorreborre commented on implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:
```diff
@@ -114,6 +118,30 @@ impl TcpSendWorker {
+        let msg_len = minicbor::len(&transport_message);
+
+        if msg_len > MAX_MESSAGE_SIZE {
+            return Err(TransportError::MessageLengthExceeded)?;
```
We could put the actual size in the error message.

etorreborre commented on implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:
```diff
@@ -114,6 +118,30 @@ impl TcpSendWorker {
+            return Err(TransportError::MessageLengthExceeded)?;
+        }
+
+        let len = 4 /* len */ + msg_len;
```
Maybe instead of `/* len */` a more explicit comment saying `We need 4 bytes to serialize the size of the message`.

etorreborre commented on implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:
```diff
@@ -133,6 +161,22 @@ impl Worker for TcpSendWorker {
             self.receiver_flow_control_id.clone(),
         ));
 
+        // First thing send our protocol version
```
It's nice to have the protocol version serialized first now.

etorreborre commented on implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:
```diff
@@ -114,6 +118,30 @@ impl TcpSendWorker {
+
+        vec[..4].copy_from_slice(&msg_len_u32.to_be_bytes());
+        minicbor::encode(&transport_message, &mut vec[4..])
+            .map_err(|_| TransportError::Encoding)?;
```
Can we keep the encoding error?

SanjoDeundiak replied:
Do you think it would be useful? Decoding errors is what can happen and having details would be useful... encoding errors are 99% bugs found during development

etorreborre replied:
In general I think it's better to not drop information, and one day missing it when debugging a production problem, at least to be able to say with 100% certainty that a bug does not come from that line.

SanjoDeundiak committed:
feat(rust): change tcp protocol serialization

SanjoDeundiak committed:
chore(rust): fix `ockam_identity` dev dependency features

SanjoDeundiak committed:
feat(rust): add secure channel padding and optimize encoding

SanjoDeundiak committed:
feat(rust): optimize cbor encoding by preallocating memory

SanjoDeundiak committed:
chore(rust): replace `default_features` with `default-features`

SanjoDeundiak committed:
chore(rust): move tracing context code to `LocalMessage`

SanjoDeundiak committed:
fix(rust): account for `minicbor` length calculation bug

polvorin committed:
feat(elixir): cbor tcp message serialization and secure channel padding

adapt tcp transport message serialization and introduce support for padding
in secure channel messages

SanjoDeundiak committed:
build: bump minicbor from 0.24.1 to 0.24.2

polvorin committed:
chore(elixir): send tcp version header immediately

do not wait for connecting side header
